File: Algorithm_Rust/examples/82. 删除排序链表中的重复元素 II.rs
use std::collections::HashMap;

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}
impl Solution {
    pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut t_node = Box::new(ListNode::new(-1));
        let new_head = &t_node;

        let mut map = HashMap::new();
        t_node.next = head;
        while let Some(node) = &t_node.next {
            *map.entry(node.val).or_insert(0) += 1;
            t_node = &t_node.next;
        }
        t_node = new_head;
        while let Some(node) = &t_node.next {
            let &mut count = map.get(&node.val).unwrap();
            while count > 0 {
                t_node = node.next;
            }
        }
        t_node.next
    }
}

fn main() {}

File: Algorithm_Rust/examples/82. 删除排序链表中的重复元素 II.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function deleteDuplicates(head: ListNode | null): ListNode | null {
  let new_head = new ListNode();
  new_head.next = head;
  let node = head;
  let map = {};
  while (node) {
    map[node.val] = (map[node.val] || 0) + 1;
    node = node.next;
  }
  console.log(map);
  let pre_node: ListNode | null = new_head;
  node = head;
  while (node) {
    console.log(node);
    let count = map[node.val];
    if (count <= 1) {
      pre_node!.next = node;
      pre_node = node;
      node = node.next;
    } else {
      console.log(count);
      while (count && node) {
        node = node?.next;
        count -= 1;
      }
      pre_node.next = node;
    }
  }

  return new_head.next;
}

export {};
